### ISSUE TITLE
Crash: SDWebImageDownloaderOperation.m:349 (-[SDWebImageDownloaderOperation URLSession:dataTask:didReceiveData:])

### ISSUE THREAD

**Issue by Jp-AFFOA** [SDWebImage/SDWebImage#2734]
### New Issue Checklist

* [ X] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ X] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ X] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | iOS
 Platform Version        | 12.2.0
 SDWebImage Version      | 5.0.2
 Integration Method      | CocoaPods
 Xcode Version           |  XCode 10.0.1
 Repro rate              | only once
 Repro with our demo prj | No
 Demo project link       | N/A

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

Unsure how to reproduce the crash. Largely passing it along, in case you might have further insights to help us chase it down.

StackTrace:
```
Fatal Exception: NSInvalidArgumentException
0  CoreFoundation                 0x1c2af7758 __exceptionPreprocess
1  libobjc.A.dylib                0x1c1cffd00 objc_exception_throw
2  CoreFoundation                 0x1c29fd8b4 -[NSCache init]
3  Foundation                     0x1c3579d28 NSAllocateMemoryPages
4  Foundation                     0x1c352c87c _NSDataCreateVMDispatchData
5  Foundation                     0x1c3467404 -[_NSPlaceholderData initWithBytes:length:copy:deallocator:]
6  SDWebImage                     0x10439a928 -[SDWebImageDownloaderOperation URLSession:dataTask:didReceiveData:] (SDWebImageDownloaderOperation.m:349)
7  SDWebImage                     0x104397af4 -[SDWebImageDownloader URLSession:dataTask:didReceiveData:] (SDWebImageDownloader.m:415)
8  CFNetwork                      0x1c3289e80 __67-[NSURLSession delegate_dataTask:didReceiveData:completionHandler:]_block_invoke_2
9  Foundation                     0x1c35736b8 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__
10 Foundation                     0x1c34790d8 -[NSBlockOperation main]
11 Foundation                     0x1c3478568 -[__NSOperationInternal _start:]
12 Foundation                     0x1c3575474 __NSOQSchedule_f
13 libdispatch.dylib              0x1c2508374 _dispatch_call_block_and_release
14 libdispatch.dylib              0x1c25098f4 _dispatch_client_callout
15 libdispatch.dylib              0x1c250c8f0 _dispatch_continuation_pop
16 libdispatch.dylib              0x1c250bfbc _dispatch_async_redirect_invoke
17 libdispatch.dylib              0x1c25188c8 _dispatch_root_queue_drain
18 libdispatch.dylib              0x1c2519148 _dispatch_worker_thread2
19 libsystem_pthread.dylib        0x1c2703b50 _pthread_wqthread
20 libsystem_pthread.dylib        0x1c2709dc4 start_wqthread
```

**Comment by dreampiggy** [comment]
Dunplicated for #2353

We have no better way to solve this. This may because OOM issue when downloading large size of image data. All the downloading-data buffer is keeping in memory.

We have a idea to solve this, by writing the data into disk during downloading. But this need many changes to solve, and this change is not compatible for [Progressive Downloading](https://sdwebimage.github.io/Enums/SDWebImageOptions.html#/c:@E@SDWebImageOptions@SDWebImageProgressiveLoad) feature.

Added this to the issue list, hope for future better solution.

**Comment by dreampiggy** [comment]
Wait. You're using the progerssive downloading feature ? I found the crash stack is in here:

```
    if (self.options & SDWebImageDownloaderProgressiveLoad) {
        // Get the image data
        NSData *imageData = [self.imageData copy]; // <-----------
}
```

This may because of too-frequency progressive decoding and copy of data. What about using [minimumProgressInterval](https://sdwebimage.github.io/Classes/SDWebImageDownloaderConfig.html#/c:objc(cs)SDWebImageDownloaderConfig(py)minimumProgressInterval) to limit the frequency ?

**Comment by Jp-AFFOA** [comment]
Thank you, @DreamPiggy. I'll give that a try. Thank you for your speed in responding.

**Comment by YueLiXing** [comment]
https://github.com/SDWebImage/SDWebImage/blob/0a3cd255a655b73fb3b3437acf2ab506b5c0c9c6/SDWebImage/Core/SDWebImageDownloaderOperation.m#L422

![image](https://user-images.githubusercontent.com/9456800/62453130-46b07600-b7a4-11e9-8a8d-fa3369a46e22.png)

**Comment by kinarobin** [comment]
@YueLiXing This is OOM issue. I strongly recommend that ：
 ```ruby
[SDImageCacheConfig defaultCacheConfig].maxMemoryCost = 300 * 1024 * 1024; //300M
```
With do that, This issue from my project had gone.

**Comment by YueLiXing** [comment]
@kinarobin  Thanks !

**Comment by YueLiXing** [comment]
[YYImage](https://github.com/ibireme/YYImage)

I've heard that people who use this library have never encountered similar problems. I suggest that authors can learn from each other.

听说用这个库的人，没遇到过类似的问题。我建议作者可以相互借鉴下

**Comment by dreampiggy** [comment]
@YueLiXing It use URLConnection. And does not have progressive downlodaing, so they don't need to `Append Data Buffer inside Memory`.

**Comment by aaparth** [comment]
> Dunplicated for #2353
> 
> We have no better way to solve this. This may because OOM issue when downloading large size of image data. All the downloading-data buffer is keeping in memory.
> 
> We have a idea to solve this, by writing the data into disk during downloading. But this need many changes to solve, and this change is not compatible for [Progressive Downloading](https://sdwebimage.github.io/Enums/SDWebImageOptions.html#/c:@E@SDWebImageOptions@SDWebImageProgressiveLoad) feature.
> 
> Added this to the issue list, hope for future better solution.

Hello @dreampiggy 

What solutions you can provide us with at the time as our app users are facing the same issue right now.
